Address the possibility of the flag introducing authentication bugs in implementations (#15)
jhoyla/draft-jhoyla-req-mtls-flag: draft-jhoyla-req-mtls-flag.md

`ClientHello` allows the server to request authentication during the initial
handshake only when it receives a hint the client will handle the request
without terminating the connection, and, if unable to authenticate, by sending
an empty `Certificate` message, per {{?RFC8446, Section 4.4.2}}.

> This mechanism is primarily intended to allow web servers to cryptographically

Reviewer: why "web" servers?

```suggestion
This mechanism is primarily intended to allow servers to cryptographically
```

authenticate web crawlers and other trusted sources of automated traffic
without prompting human users to present a certificate.

The traffic generated by these sources represents a significant and ever
increasing fraction of the overall traffic to a given web server. Many web
servers handle this traffic differently than they handle traffic from web
browsers operated by human users. On the one hand, human users are much more
latency sensitive than web crawlers. On the other hand, it may be prudent to
treat requests from an unknown client with more scrutiny than requests from a
well-known web-crawler whose behaviour is relatively predictable.

It is therefore common for web servers to attempt to identify trusted sources

> of traffic so that they can be distinguished between one another and,
> especially, from human users. This dynamic creates an incentive for attackers

Reviewer: slightly changing the wording to make it clear we want to identify each crawler individually, and distinguish them from human users

```suggestion
of traffic so that they can be distinguished between one another, as well as from human users. This dynamic creates an incentive for attackers
```

to impersonate trusted sources of automated traffic to the server: to confer
on themselves whatever accommodations are conferred on trusted sources.
Unfortunately, many of the features used for self-identification are easily
spoofed {{?I-D.draft-meunier-web-bot-auth-architecture}}.

> Authentication of the TLS client is sufficient to thwart such attacks, and

Reviewer: clearly mentioning the assumption that crawlers leverage a TLS client

```suggestion
In its current state, most web crawlers scrape the web via a TLS client.
Authentication of the TLS client is sufficient to thwart impersonation attacks, and
```

while there are other approaches to solving this problem (see
{{alternative-approaches}}), authentication at the TLS layer has several
desirable features for this use case. [[TODO: Address issue \#12 and add a
reference here.]]

> The mechanism specified in this document may be useful for other use cases.

Reviewer: ```suggestion
The mechanism specified in this document could be applied to use cases beyond crawler identification.
```

> However, it may entail security, privacy, or operational considerations that

Reviewer: ```suggestion
However, doing so may entail security, privacy, or operational considerations that
```

are relevant to these use cases but are not directly addressed here.

> In particular, the `request_client_auth` flag is not intended to be used by web

Reviewer: request_client_auth has not been introduced yet.
this may also be a reference to the specific definition section via `{#id}` and `{{id}}`

```suggestion
In particular, `request_client_auth` flag, the TLS flag defined in this document, is not intended to be used by web
```

`ClientHello` allows the server to request authentication during the initial
handshake only when it receives a hint the client will handle the request
without terminating the connection, and, if unable to authenticate, by sending
an empty `Certificate` message, per {{?RFC8446, Section 4.4.2}}.

This mechanism is primarily intended to allow servers to cryptographically
authenticate web crawlers and other trusted sources of automated traffic
without prompting human users to present a certificate.

The traffic generated by these sources represents a significant and ever
increasing fraction of the overall traffic to a given web server. Many web
servers handle this traffic differently than they handle traffic from web
browsers operated by human users. On the one hand, human users are much more
latency sensitive than web crawlers. On the other hand, it may be prudent to
treat requests from an unknown client with more scrutiny than requests from a
well-known web-crawler whose behaviour is relatively predictable.

It is therefore common for web servers to attempt to identify trusted sources

> of traffic so that they can be distinguished between one another, as well as

Reviewer 2: ```suggestion
of traffic so that the server can distinguish between them, as well as differentiate them
```

from human users. This dynamic creates an incentive for attackers to
impersonate trusted sources of automated traffic to the server: to confer on
themselves whatever accommodations are conferred on trusted sources.
Unfortunately, many of the features used for self-identification are easily
spoofed {{?I-D.draft-meunier-web-bot-auth-architecture}}.

Reaching the server normally requires first establishing a TLS {{?RFC8446}}
connection with it. Authentication of the TLS client is sufficient to thwart
impersonation attacks, and while there are other approaches to solving this

> problem (see {{alternative-approaches}}), authentication at the TLS layer has

Reviewer 2: Please could you add an anchor to the Alternative Approaches section?

several desirable features for this use case. [[TODO: Address issue \#12 and
add a reference here.]]

The mechanism specified in this document may be used for other purposes, but
doing so may entail security, privacy, or operational considerations that are
relevant to these use cases but are not directly addressed here.

In particular, the `request_client_auth` flag is not intended to be used by web
browsers or other clients operated directly by human users. Clients that set

> this flag should only do so when a client certificate is directly available to

Reviewer 2: Please put normative SHOULD and MUST NOT for these later in the draft.

> # Security Considerations

Reviewer: would it be relevant to point to security consideration section of TLS flags ? https://datatracker.ietf.org/doc/html/draft-ietf-tls-tlsflags-15#name-security-considerations

Reviewer: would it make sense to update the wording ofn tls flag draft then?

Reviewer 2: Although now I think on it, maybe we should just say "ignore" because otherwise we're giving servers a way to detect whether a client understands this draft. A client that ignores the incorrect flag doesn't, one that aborts does.

On its own, this flag should have no effect on the security of TLS, as the
server may always send a `CertificateRequest` message during the handshake.
This flag merely provides a hint that the client will handle the request
gracefully. Because the server acknowledges the flag in the
`CertificateRequest` the client can always be sure whether a
`CertificateRequest` was triggered by `request_client_auth` or not.

However, the use case for which the flag is enabled may entail significant
changes to how the client selects a certificate and what trust anchors the
server uses to validate it. The use case may also impact how the server handles
application data transmitted over the established connection.

> In the case of identifying web crawlers (see {{introduction}}), a web server

Reviewer: ```suggestion
For the identifying web crawlers use case mentioned in {{introduction}}, a web server
```

Reviewer: also: you should likely define `{#introduction}`

would largely handle traffic from a web crawler as if it were any
unauthenticated client. If for instance the server already requests a
certificate to grant certain clients access to some resource, the server must take
care to ensure this access is not also granted to the web crawler.

When the server accepts the `request_client_auth` flag, it SHOULD verify the
certificate against a set of trust anchors that is disjoint from the set of

> trust anchors it uses when the flag is ignored or rejected.

Reviewer: not sure this is the right action. my understanding is if a server accepts the flag but does not verify the certificate, there's a high change of misimplementation, and we should recommend to terminate the connection

```suggestion
trust anchors it uses when the flag is ignored or rejected. If it does
not, it is RECOMMENDED to terminate the connection.
```
